File: backend/src/shared/services/r2.service.ts
import {
  DeleteObjectCommand,
  GetObjectCommand,
  S3Client,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import crypto from "crypto";
import logger from "../../utils/logger";

const credentials = {
  accessKeyId: process.env.R2_ACCESS_KEY_ID!,
  secretAccessKey: process.env.R2_SECRET_ACCESS_KEY!,
};

const r2Client = new S3Client({
  region: "us-east-1",
  endpoint: process.env.R2_ENDPOINT!,
  credentials,
  forcePathStyle: true,
});

/**
 * Separate client used only to sign URLs.
 *
 * A signature covers the host, so a URL signed against `R2_ENDPOINT` is only
 * valid at that host — fine for Cloudflare R2, whose S3 endpoint is publicly
 * reachable, but useless when the bucket is reached over a private network
 * (a self-hosted MinIO on `http://minio:9000`, say). `R2_PUBLIC_ENDPOINT` lets
 * such a deployment sign against the host the browser actually talks to, and
 * falls back to `R2_ENDPOINT` so existing setups keep working untouched.
 */
const signingClient = process.env.R2_PUBLIC_ENDPOINT
  ? new S3Client({
      region: "us-east-1",
      endpoint: process.env.R2_PUBLIC_ENDPOINT,
      credentials,
      forcePathStyle: true,
    })
  : r2Client;

/** How long a signed resume link stays valid. */
const SIGNED_URL_TTL_SECONDS = Number(
  process.env.SIGNED_URL_TTL_SECONDS ?? 900,
);

// Server-controlled extension per content type. We never trust the client
// filename for the stored key (an attacker could upload a `.html`/`.svg`).
const EXT_BY_MIME: Record<string, string> = {
  "application/pdf": ".pdf",
  "image/png": ".png",
  "image/jpeg": ".jpg",
  "image/webp": ".webp",
  "image/svg+xml": ".svg",
};

export const r2Service = {
  async uploadFile(
    file: Express.Multer.File,
    folder: "resumes" | "logos" = "resumes",
  ): Promise<string> {
    const fileExt = EXT_BY_MIME[file.mimetype] ?? "";
    const fileName = `${folder}/${crypto.randomUUID()}${fileExt}`;

    // Logos are raster images (png/jpeg/webp — never svg) meant to render
    // inline in the UI. Everything else (resumes, and any svg) is forced to
    // download so an uploaded file can never execute as HTML/SVG when its
    // URL is opened directly in a browser.
    const isInlineableLogo =
      folder === "logos" &&
      ["image/png", "image/jpeg", "image/webp"].includes(file.mimetype);

    const command = new PutObjectCommand({
      Bucket: process.env.R2_BUCKET_NAME!,
      Key: fileName,
      Body: file.buffer,
      ContentType: file.mimetype,
      ContentDisposition: isInlineableLogo ? "inline" : "attachment",
    });

    try {
      await r2Client.send(command);
    } catch (error) {
      logger.error(
        `Attempting upload to Bucket: ${process.env.R2_BUCKET_NAME}`,
      );
      throw error;
    }

    const publicUrl = process.env.R2_PUBLIC_URL?.endsWith("/")
      ? process.env.R2_PUBLIC_URL.slice(0, -1)
      : process.env.R2_PUBLIC_URL;

    return `${publicUrl}/${fileName}`;
  },

  extractKeyFromUrl(fileUrl: string): string | null {
    if (!fileUrl) return null;

    const base = process.env.R2_PUBLIC_URL?.endsWith("/")
      ? process.env.R2_PUBLIC_URL.slice(0, -1)
      : process.env.R2_PUBLIC_URL;

    if (!base) return null;
    if (!fileUrl.startsWith(`${base}/`)) return null;

    return fileUrl.replace(`${base}/`, "");
  },

  /**
   * Turns a stored file URL into a short-lived signed one.
   *
   * Resumes are personal data, so `resumes/` is not anonymously readable and a
   * stored URL alone opens nothing. Callers hand the signed URL to an already
   * authorised viewer; it expires on its own, which a permanent public link
   * never does.
   *
   * Returns the URL unchanged when it points outside our bucket, and null for
   * null input, so callers can map over optional values without branching.
   */
  async signUrl(fileUrl: string | null): Promise<string | null> {
    if (!fileUrl) return null;

    const key = this.extractKeyFromUrl(fileUrl);
    if (!key) return fileUrl;

    try {
      return await getSignedUrl(
        signingClient,
        new GetObjectCommand({
          Bucket: process.env.R2_BUCKET_NAME!,
          Key: key,
        }),
        { expiresIn: SIGNED_URL_TTL_SECONDS },
      );
    } catch (error) {
      // A viewer seeing no resume is better than a 500 on the whole record.
      logger.error(`Failed to sign URL for key ${key}: ${error}`);
      return null;
    }
  },

  async deleteByUrl(fileUrl: string): Promise<void> {
    const key = this.extractKeyFromUrl(fileUrl);
    if (!key) return;

    const command = new DeleteObjectCommand({
      Bucket: process.env.R2_BUCKET_NAME!,
      Key: key,
    });

    await r2Client.send(command);
  },
};

File: backend-py/app/shared/services/r2_service.py
"""Cloudflare R2 storage (S3-compatible), equivalent to
backend/src/shared/services/r2.service.ts.

Two clients on purpose: `_r2_client` for uploads/deletes (R2_ENDPOINT), and
`_signing_client` only to presign URLs (R2_PUBLIC_ENDPOINT if set, else
falls back to R2_ENDPOINT) - a presigned URL's signature is bound to the
host it was signed against, so a self-hosted deployment behind a private
network needs a distinct public-facing host to sign against.
"""

from __future__ import annotations

import uuid
from typing import Literal
from urllib.parse import urlparse

import boto3
from botocore.client import Config

from app.logging import get_logger
from app.settings import settings

logger = get_logger(__name__)

Folder = Literal["resumes", "logos"]

# Whitelist mapping mimetype -> extension for server-controlled filenames -
# never trust a client-supplied filename (prevents e.g. a .html/.svg upload
# disguised as something else).
_EXT_BY_MIME = {
    "application/pdf": ".pdf",
    "image/png": ".png",
    "image/jpeg": ".jpg",
    "image/webp": ".webp",
    "image/svg+xml": ".svg",
}

_INLINE_SAFE_MIME_TYPES_FOR_LOGOS = {"image/png", "image/jpeg", "image/webp"}


def _build_client(endpoint: str):
    return boto3.client(
        "s3",
        endpoint_url=endpoint,
        aws_access_key_id=settings.r2_access_key_id,
        aws_secret_access_key=settings.r2_secret_access_key,
        region_name="us-east-1",  # arbitrary/ignored by R2
        config=Config(
            signature_version="s3v4",
            s3={"addressing_style": "path"},
            connect_timeout=5,
            read_timeout=10,
            retries={"max_attempts": 2},
        ),
    )


_r2_client = _build_client(settings.r2_endpoint)
_signing_client = _build_client(settings.r2_public_endpoint or settings.r2_endpoint)


def _public_url(file_name: str) -> str:
    return f"{settings.r2_public_url.rstrip('/')}/{file_name}"


def upload_file(*, content: bytes, content_type: str, folder: Folder = "resumes") -> str:
    ext = _EXT_BY_MIME.get(content_type, "")
    file_name = f"{folder}/{uuid.uuid4().hex}{ext}"

    is_inline_safe_logo = folder == "logos" and content_type in _INLINE_SAFE_MIME_TYPES_FOR_LOGOS
    disposition = "inline" if is_inline_safe_logo else "attachment"

    _r2_client.put_object(
        Bucket=settings.r2_bucket_name,
        Key=file_name,
        Body=content,
        ContentType=content_type,
        ContentDisposition=disposition,
    )
    return _public_url(file_name)


def download_file(key: str) -> bytes:
    response = _r2_client.get_object(Bucket=settings.r2_bucket_name, Key=key)
    return response["Body"].read()


def extract_key_from_url(file_url: str) -> str | None:
    base = settings.r2_public_url.rstrip("/")
    if not file_url.startswith(base):
        return None
    key = file_url[len(base) :].lstrip("/")
    return urlparse(f"https://x/{key}").path.lstrip("/")


def sign_url(file_url: str | None) -> str | None:
    """Converts a stored (unsigned) public-style URL into a short-lived
    presigned GET URL. Never raises: a signing failure logs and returns
    None so callers can degrade to "no resume" instead of a 500."""
    if file_url is None:
        return None

    key = extract_key_from_url(file_url)
    if key is None:
        return file_url  # points outside our bucket - pass through unchanged

    try:
        return _signing_client.generate_presigned_url(
            "get_object",
            Params={"Bucket": settings.r2_bucket_name, "Key": key},
            ExpiresIn=settings.signed_url_ttl_seconds,
        )
    except Exception:  # noqa: BLE001
        logger.exception("failed to sign url for key=%s", key)
        return None


def delete_by_url(file_url: str | None) -> None:
    if file_url is None:
        return
    key = extract_key_from_url(file_url)
    if key is None:
        return
    _r2_client.delete_object(Bucket=settings.r2_bucket_name, Key=key)
